### PR TITLE
Fortran header

### PR DESCRIPTION
**Pull request by chiaraMaHe**
The header with the first basic functions for the Fortran interface were added and it was added to the Makefile.

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: src/t8_fortran_interface/t8_fortran_interface.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
+/*
+  This file is part of t8code.
+  t8code is a C library to manage a collection (a forest) of multiple
+  connected adaptive space-trees of general element classes in parallel.
+
+  Copyright (C) 2015 the developers
```
Comment (sandro-elsweijer) [inline src/t8_fortran_interface/t8_fortran_interface.h:6]:
```suggestion
  Copyright (C) 2024 the developers

```

**Comment by sandro-elsweijer** [comment]
I do not know how we will finally split the C and CPP interface, but we will need a new folder structure for this. Maybe we can do it like here and have a separate API folder for every API?
https://gitlab.onelab.info/gmsh/gmsh
Then we could also add the Julia api as an optional submodule. We could do the same with the Fortran interface.

File: api/Makefile.am [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+if T8_ENABLE_FORTRAN
+fortraninterfaceincludedir = $(includedir)/t8_fortran_interface
+
+libt8_installed_headers_fortran_interface = \
+  api/t8_fortran_interface/t8_fortran_interface.h
+
+dist_fortraninterfaceinclude_HEADERS = $(libt8_installed_headers_fortran_interface)
+
+AM_CPPFLAGS += -I@top_srcdir@/api
+endif
```
Comment (sandro-elsweijer) [inline api/Makefile.am:14]:
```suggestion
endif

```

File: api/t8_fortran_interface/t8_fortran_interface.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+void
+t8_fortran_finalize ();
+
+/** Commit cmesh. This wraps cmesh_commit in order to use the dereferenced communicator.
+*/
+void
+t8_fortran_cmesh_commit (t8_cmesh_t cmesh, sc_MPI_Comm *comm);
+
+/** This function calls t8_cmesh_set_join_by_vertices with connectivity = NULL.
+*/
```
Comment (sandro-elsweijer) [inline api/t8_fortran_interface/t8_fortran_interface.h:67]: Please use docstrings here

File: api/t8_fortran_interface/t8_fortran_interface.h [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+/** Free the memory of a C MPI Communicator pointer that was created
+ *  with \ref t8_fortran_MPI_Comm_new.
+ * \param [in] Ccomm  Pointer to a C MPI communicator.
+ */
+void
+t8_fortran_MPI_Comm_delete (sc_MPI_Comm *Ccomm);
+
+t8_cmesh_t
+t8_cmesh_new_periodic_tri_wrap (sc_MPI_Comm *Ccomm);
```
Comment (sandro-elsweijer) [inline api/t8_fortran_interface/t8_fortran_interface.h:98]: docstring

File: api/t8_fortran_interface/t8_fortran_interface.h [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+ *                            \a scheme and refinement level \a level.
+ */
+t8_forest_t
+t8_forest_new_uniform_default (t8_cmesh_t cmesh, int level, int do_face_ghost, sc_MPI_Comm *comm);
+
+t8_forest_t
+t8_forest_adapt_by_coordinates (t8_forest_t forest, int recursive, t8_fortran_adapt_coordinate_callback callback);
+
+void
+t8_global_productionf_noargs (const char *string);
```
Comment (sandro-elsweijer) [inline api/t8_fortran_interface/t8_fortran_interface.h:117]: docstrings

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]
I am not able to test the functionality, since I do not have a clue of Fortran, but it looks alright.
Btw, how does this work? These are just declarations, where are the actual definitions? And can Fortran just link to a c header file, or where are the Fortran files?

File: api/t8_fortran_interface/t8_fortran_interface.h [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+/** Finalize sc. This wraps sc_finalize in order to have consistent
+ * naming with t8_fortran_init_all.
+ */
+void
+t8_fortran_finalize ();
+
+/** Commit cmesh. This wraps cmesh_commit in order to use the dereferenced communicator.
+ * \param [in] Ccomm  Pointer to a C MPI communicator.
+*/
```
Comment (sandro-elsweijer) [inline api/t8_fortran_interface/t8_fortran_interface.h:62]: The docstring does not match

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]
Last comment :)

**Comment by chiaraMaHe** [comment]
> I am not able to test the functionality, since I do not have a clue of Fortran, but it looks alright. Btw, how does this work? These are just declarations, where are the actual definitions? And can Fortran just link to a c header file, or where are the Fortran files?

This is only the c header file, that will be imported into a c file, which defines functions that are needed by the f90 file. It makes it easier because you sometimes need to pass the pointer instead of the value (Ccomm). The c-file and the f90 file will follow in the upcoming PRs

File: api/t8_fortran_interface/t8_fortran_interface.h [hunk 7]
```diff
@@ -0,0 +1,149 @@
+/*
+  This file is part of t8code.
+  t8code is a C library to manage a collection (a forest) of multiple
+  connected adaptive space-trees of general element classes in parallel.
+
+  Copyright (C) 2024 the developers
+
```
Comment (sandro-elsweijer) [inline api/t8_fortran_interface/t8_fortran_interface.h:7]:
```suggestion

```

**Review by sandro-elsweijer — APPROVED** [review]

**Review by sandro-elsweijer — APPROVED** [review]
Almost

**Pull request merged**